api/clients/v2/accountant.go

```diff
@@ -5,115 +5,106 @@ import (
+		accountID:    accountID,
+		reservations: reservations,
+		onDemand:     onDemand,
+		paymentVaultParams: &meterer.PaymentVaultParams{
+			QuorumPaymentConfigs:  make(map[uint8]*core.PaymentQuorumConfig),
+			QuorumProtocolConfigs: make(map[uint8]*core.PaymentQuorumProtocolConfig),
+			OnDemandQuorumNumbers: make([]uint8, 0),
+		},
 		periodRecords:     periodRecords,
 		cumulativePayment: big.NewInt(0),
 	}
-	// TODO: add a routine to refresh the on-chain state occasionally?
-	return &a
 }
 
-// reservationUsage attempts to use the reservation for the given request.
-func (a *Accountant) reservationUsage(
-	symbolUsage uint64,
-	quorumNumbers []uint8,
-	timestamp int64) error {
-	if err := meterer.ValidateQuorum(quorumNumbers, a.reservation.QuorumNumbers); err != nil {
+// ReservationUsage attempts to use the reservation for the requested quorums; if any quorum fails to use the reservation, the entire operation is rolled back.
+func (a *Accountant) reservationUsage(numSymbols uint64, quorumNumbers []core.QuorumID, timestamp int64) error {
+	// The two timestamps are the same for the accountant client for validating the reservation period; the second timestamp is the received at time for the server
+	if err := meterer.ValidateReservations(a.reservations, a.paymentVaultParams.QuorumProtocolConfigs, quorumNumbers, timestamp, time.Unix(0, timestamp)); err != nil {
 		return err
 	}
-	if !a.reservation.IsActiveByNanosecond(timestamp) {
-		return fmt.Errorf("reservation is not active at timestamp %d", timestamp)
-	}
-
-	reservationWindow := a.reservationWindow
-	currentReservationPeriod := meterer.GetReservationPeriodByNanosecond(timestamp, reservationWindow)
 
 	a.periodRecordsLock.Lock()
 	defer a.periodRecordsLock.Unlock()
-	relativePeriodRecord := a.getOrRefreshRelativePeriodRecord(currentReservationPeriod, reservationWindow)
-	relativePeriodRecord.Usage += symbolUsage
-
-	// Check if we can use the reservation within the bin limit
-	binLimit := meterer.GetReservationBinLimit(a.reservation, a.reservationWindow)
-	if relativePeriodRecord.Usage <= binLimit {
-		return nil
-	}
-
-	overflowPeriodRecord := a.getOrRefreshRelativePeriodRecord(meterer.GetOverflowPeriod(currentReservationPeriod, reservationWindow), reservationWindow)
-	// Allow one overflow when the overflow bin is empty, the current usage and new length are both less than the limit
-	if overflowPeriodRecord.Usage == 0 && relativePeriodRecord.Usage-symbolUsage < binLimit && symbolUsage <= binLimit {
-		overflowPeriodRecord.Usage += relativePeriodRecord.Usage - binLimit
-		return nil
+	// deep copy of periodRecords for rollback in case of errors
+	originalPeriodRecords := a.periodRecords.DeepCopy()
+
+	for _, quorumNumber := range quorumNumbers {
+		reservation, exists := a.reservations[quorumNumber]
+		if !exists {
+			// this case should never happen because ValidateReservations should have already checked this; handle it just in case
+			a.periodRecords = originalPeriodRecords
```

Comment:
Instead of working on the original periodRecords and then rolling back on every error path (super bug prone - easy to forget one while refactoring), better to instead update the copy and at the very bottom commit it (only commit in the single happy return path)
```
periodRecordsCopy := a.periodRecords.DeepCopy()
// modify periodRecordsCopy
// ....
// at the bottom of the function...
a.periodRecords = periodRecordsCopy
return nil
```

Reply:
updated in [commit](https://github.com/Layr-Labs/eigenda/pull/1664/commits/783ec1c7184a942bdfde1ef6ad4ffc611d96c776)